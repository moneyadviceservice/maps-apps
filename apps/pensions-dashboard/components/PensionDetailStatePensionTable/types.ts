export type Row = {
  title: string;
  monthlyAmount: string;
  yearlyAmount: string;
  payableDate: string;
};
