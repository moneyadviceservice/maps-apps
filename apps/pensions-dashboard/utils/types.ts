export type ErrorResponse = {
  error: string;
  message: string;
};

export type PensionData = {
  pensionPolicies: PensionPolicy[];
  peiInformation: PeiInformationModel;
  pensionsDataRetrievalComplete: boolean;
};

type PensionPolicy = {
  pensionArrangements: PensionArrangement[];
};

export type PensionArrangement = {
  externalPensionPolicyId: string;
  externalAssetId: string;
  matchType: 'DEFN' | 'POSS' | 'CONT' | 'NEW' | 'SYS';
  schemeName: string;
  alternateSchemeNames?: AlternateScheme[];
  contactReference: string;
  pensionType: 'AVC' | 'CB' | 'CDC' | 'DB' | 'DC' | 'HYB' | 'SP';
  pensionOrigin?: 'A' | 'PC' | 'PM' | 'PT' | 'WC' | 'WM' | 'WT';
  pensionStatus?: 'A' | 'I' | 'IPPF' | 'IWU';
  startDate?: string;
  retirementDate?: string;
  dateOfBirth?: string;
  statePensionMessageEng?: string;
  statePensionMessageWelsh?: string;
  contributionsFromMultipleEmployers: boolean;
  pensionAdministrator: PensionAdministrator;
  employmentMembershipPeriods?: EmploymentMembershipPeriod[];
  benefitIllustrations?: BenefitIllustration[];
  additionalDataSources?: AdditionalDataSource[];
};

type AlternateScheme = {
  name: string;
  alternateNameType: 'FOR' | 'OTH';
};

type PensionAdministrator = {
  name: string;
  contactMethods: ContactMethod[];
};

type ContactMethod = {
  preferred: boolean;
  contactMethodDetails: EmailAddress | PhoneNumber | PostalAddress | Website;
};

type EmploymentMembershipPeriod = {
  employerName: string;
  employerStatus: 'C' | 'H';
  membershipStartDate: string;
  membershipEndDate: string;
};

type BenefitIllustration = {
  illustrationComponents: BenefitIllustrationComponent[];
  illustrationDate: string;
};

type AdditionalDataSource = {
  url: string;
  informationType: 'C_AND_C' | 'SIP' | 'IMP' | 'ANR' | 'SP';
};

type PeiInformationModel = {
  peiRetrievalComplete: boolean;
  peiData: PeiData[];
};

type PeiData = {
  pei: string;
  description: string;
  retrievalStatus:
    | 'NEW'
    | 'RETRIEVAL_REQUESTED'
    | 'RETRIEVAL_COMPLETE'
    | 'RETRIEVAL_FAILED';
};

type EmailAddress = {
  email: string;
};

type PhoneNumber = {
  number: string;
  usage: ('A' | 'M' | 'N' | 'S' | 'W')[];
};

type PostalAddress = {
  postalName: string;
  line1: string;
  line2?: string;
  line3?: string;
  line4?: string;
  line5?: string;
  postcode: string;
  countryCode: string;
};

type Website = {
  url: string;
};

type BenefitIllustrationComponent = {
  illustrationType: 'ERI' | 'AP';
  unavailableReason?:
    | 'ANO'
    | 'DB'
    | 'DBC'
    | 'DCC'
    | 'DCHA'
    | 'DCHP'
    | 'DCSM'
    | 'MEM'
    | 'NEW'
    | 'NET'
    | 'PPF'
    | 'TRN'
    | 'WU';
  benefitType: 'AVC' | 'CBL' | 'CBS' | 'CDI' | 'CDL' | 'DB' | 'DBL' | 'DC';
  calculationMethod: 'BS' | 'CBI' | 'SMPI';
  payableDetails: PayableDetails;
  dcPot?: number;
  survivorBenefit?: boolean;
  safeguardedBenefit?: boolean;
  illustratonWarnings?: (
    | 'AVC'
    | 'CUR'
    | 'DEF'
    | 'FAS'
    | 'PEO'
    | 'PNR'
    | 'PSO'
    | 'SPA'
    | 'TVI'
    | 'INP'
  )[];
};

type PayableDetails =
  | RecurringIncomeDetails
  | LumpSumDetails
  | AmountNotProvidedDetails;

type RecurringIncomeDetails = {
  amountType: 'INC' | 'INCL' | 'INCN';
  annualAmount: number;
  monthlyAmount: number;
  lastPaymentDate?: string;
  payableDate: string;
  increasing: boolean;
};

type LumpSumDetails = {
  amountType: 'CSH' | 'CSHL' | 'CSHN';
  amount: number;
  payableDate: string;
};

type AmountNotProvidedDetails = {
  reason: 'SML';
  payableDate: string;
  lastPaymentDate?: string;
};

export type PensionsOverviewModel = {
  totalPensions: number;
  confirmedPensions: PensionArrangement[];
  unconfirmedPensions: PensionArrangement[];
  incompletePensions: PensionArrangement[];
};
