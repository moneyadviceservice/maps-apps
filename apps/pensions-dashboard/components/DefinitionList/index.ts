export * from './DefinitionDescription';
export * from './DefinitionList';
export * from './DefinitionTerm';
