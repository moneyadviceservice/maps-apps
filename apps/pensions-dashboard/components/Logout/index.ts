export * from './LogoutDialog';
export * from './LogoutLink';
export * from './LogoutLinkText';
export * from './useLogoutHandler';
