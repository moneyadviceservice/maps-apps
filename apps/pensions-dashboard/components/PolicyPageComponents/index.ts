export * from './List';
export * from './Section';
