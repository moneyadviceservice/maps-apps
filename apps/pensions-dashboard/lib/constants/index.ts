import Cookies from 'cookies';

export const STATE_RETIREMENT_AGE = 67;

export const NO_DATA = '--';

export const NUMBER_WORDS = [
  { en: 'zero', cy: 'sero' },
  { en: 'one', cy: 'un' },
  { en: 'two', cy: 'dau' },
  { en: 'three', cy: 'tri' },
  { en: 'four', cy: 'pedwar' },
  { en: 'five', cy: 'pump' },
  { en: 'six', cy: 'chwech' },
  { en: 'seven', cy: 'saith' },
  { en: 'eight', cy: 'wyth' },
  { en: 'nine', cy: 'naw' },
  { en: 'ten', cy: 'deg' },
  { en: 'eleven', cy: 'un ar ddeg' },
  { en: 'twelve', cy: 'deuddeg' },
];

export const PROTOCOL =
  process?.env.NODE_ENV === 'development' ? 'http://' : 'https://';

export const COOKIE_OPTIONS = {
  path: '/',
  httpOnly: true,
  sameSite: 'lax',
} as Cookies.SetOption;

export enum PensionsCardType {
  CONFIRMED = 'CONFIRMED',
  CONFIRMED_NO_INCOME = 'CONFIRMED_NO_INCOME',
  PENDING = 'PENDING',
}

export enum MatchType {
  DEFN = 'DEFN',
  POSS = 'POSS',
  CONT = 'CONT',
  NEW = 'NEW',
  SYS = 'SYS',
}

export enum PensionType {
  AVC = 'AVC',
  CB = 'CB',
  CDC = 'CDC',
  DB = 'DB',
  DC = 'DC',
  HYB = 'HYB',
  SP = 'SP',
}

export enum PensionOrigin {
  A = 'A',
  PC = 'PC',
  PM = 'PM',
  PT = 'PT',
  WC = 'WC',
  WM = 'WM',
  WT = 'WT',
}

export enum PensionStatus {
  A = 'A',
  I = 'I',
  IPPF = 'IPPF',
  IWU = 'IWU',
}

export enum AlternateNameType {
  FOR = 'FOR',
  OTH = 'OTH',
}

export enum EmployerStatus {
  C = 'C',
  H = 'H',
}

export enum InformationType {
  C_AND_C = 'C_AND_C',
  SIP = 'SIP',
  IMP = 'IMP',
  ANR = 'ANR',
  SP = 'SP',
}

export enum RetrievalStatus {
  NEW = 'NEW',
  RETRIEVAL_REQUESTED = 'RETRIEVAL_REQUESTED',
  RETRIEVAL_COMPLETE = 'RETRIEVAL_COMPLETE',
  RETRIEVAL_FAILED = 'RETRIEVAL_FAILED',
}

export enum UsageType {
  A = 'A',
  M = 'M',
  N = 'N',
  S = 'S',
  W = 'W',
}

export enum IllustrationType {
  ERI = 'ERI',
  AP = 'AP',
  DB = 'DB',
}

export enum BenefitType {
  AVC = 'AVC',
  CBL = 'CBL',
  CBS = 'CBS',
  CDI = 'CDI',
  CDL = 'CDL',
  DB = 'DB',
  DBL = 'DBL',
  DC = 'DC',
  SP = 'SP',
}

export enum UnavailableReason {
  ANO = 'ANO',
  DB = 'DB',
  DBC = 'DBC',
  DCC = 'DCC',
  DCHA = 'DCHA',
  DCHP = 'DCHP',
  DCSM = 'DCSM',
  MEM = 'MEM',
  NEW = 'NEW',
  NET = 'NET',
  PPF = 'PPF',
  TRN = 'TRN',
  WU = 'WU',
}

export enum IllustrationWarning {
  AVC = 'AVC',
  CUR = 'CUR',
  DEF = 'DEF',
  FAS = 'FAS',
  PEO = 'PEO',
  PNR = 'PNR',
  PSO = 'PSO',
  SPA = 'SPA',
  TVI = 'TVI',
  INP = 'INP',
}

export enum CalculationMethod {
  BS = 'BS',
  CBI = 'CBI',
  SMPI = 'SMPI',
}

export enum LumpSumAmountType {
  CSH = 'CSH',
  CSHL = 'CSHL',
  CSHN = 'CSHN',
}

export enum RecurringAmountType {
  INC = 'INC',
  INCL = 'INCL',
  INCN = 'INCN',
}

export enum AmountNotProvidedReason {
  SML = 'SML',
}

export enum ContactMethods {
  TELEPHONE = 'number',
  EMAIL = 'email',
  POSTAL_NAME = 'postalName',
  WEB_CONTACTS = 'url',
}

export const BACK_LINKS: { [key in string]: string } = {
  CONFIRMED: '/your-pension-breakdown',
  INCOMPLETE: '/pending-pensions',
  UNCONFIRMED: '/pensions-that-need-action',
  '': '/your-pension-search-results',
};

// API response messages
export const REQUEST_ABANDONED = 'Network request was abandoned';
export const RESPONSE_NOT_OK = 'Network response was not ok';
export const RESPONSE_NOT_ACCEPTED = 'Network response was not accepted';
export const REQUEST_FAILED = 'Network request failed';
export const DATA_NOT_FOUND = 'Data not found';

export const DEFAULT_LOAD_TIME = 80;

export enum PensionGroup {
  GREEN = 'green',
  GREEN_NO_INCOME = 'green_no_income',
  RED = 'red',
  YELLOW = 'yellow',
}
