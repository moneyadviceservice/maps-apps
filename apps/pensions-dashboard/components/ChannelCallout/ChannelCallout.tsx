import { Callout, CalloutVariant } from '@maps-react/common/components/Callout';
import { Heading } from '@maps-react/common/components/Heading';
import { Icon, IconType } from '@maps-react/common/components/Icon';
import { Link } from '@maps-react/common/components/Link';
import { Paragraph } from '@maps-react/common/components/Paragraph';
import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionArrangement } from '../../lib/types';
import { PensionsList } from '../PensionsList';

type ChannelCalloutProps = {
  data: PensionArrangement[];
  variant: CalloutVariant;
  title: string;
  description?: string;
  icon: IconType;
  iconClasses?: string;
  link: string;
  linkText: string;
};

export const ChannelCallout = ({
  data,
  variant,
  title,
  description,
  icon,
  iconClasses,
  link,
  linkText,
}: ChannelCalloutProps) => {
  const { locale } = useTranslation();

  return (
    <Callout variant={variant} className="p-8 pt-6 mb-12">
      <Heading level="h4" className="mb-3">
        {title}
      </Heading>

      {description && <Paragraph>{description}</Paragraph>}

      <PensionsList
        pensions={data}
        icon={<Icon type={icon} data-testid="icon" className={iconClasses} />}
      />

      <Link
        asButtonVariant="primary"
        href={`/${locale}${link}`}
        className="mt-6 mb-2"
      >
        {linkText}
      </Link>
    </Callout>
  );
};
