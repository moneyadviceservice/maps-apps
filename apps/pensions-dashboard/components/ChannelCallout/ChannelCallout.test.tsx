import { render } from '@testing-library/react';

import { CalloutVariant } from '@maps-react/common/components/Callout';
import { IconType } from '@maps-react/common/components/Icon';
import useTranslation from '@maps-react/hooks/useTranslation';

import { mockPensionsData } from '../../lib/mocks';
import { PensionArrangement } from '../../lib/types';
import { ChannelCallout } from './ChannelCallout';

import '@testing-library/jest-dom/extend-expect';

jest.mock('@maps-react/hooks/useTranslation');

const mockUseTranslation = useTranslation as jest.Mock;

describe('ChannelCallout', () => {
  beforeEach(() => {
    mockUseTranslation.mockReturnValue({
      locale: 'en',
      t: (key: string) => key,
      tList: (key: string) => key,
    });
  });

  const mockData = {
    data: mockPensionsData.pensionPolicies[0]
      .pensionArrangements as PensionArrangement[],
    variant: CalloutVariant.POSITIVE,
    title: 'Test Title',
    description: 'Test Description',
    icon: IconType.CLOCK,
    iconClasses: 'test-icon-class',
    link: '/test-link',
    linkText: 'Test Link Text',
  };

  it('renders correctly', () => {
    const { container } = render(<ChannelCallout {...mockData} />);
    expect(container).toMatchSnapshot();
  });

  it.each([
    ['renders the title', 'Test Title'],
    ['renders the description if provided', 'Test Description'],
  ])('%s', (_, text) => {
    const { getByText } = render(<ChannelCallout {...mockData} />);
    expect(getByText(text)).toBeInTheDocument();
  });

  it('does not render the description if not provided', () => {
    const { queryByText } = render(
      <ChannelCallout {...mockData} description={undefined} />,
    );
    expect(queryByText('Test Description')).not.toBeInTheDocument();
  });

  it('renders the link with correct text and href', () => {
    const { getByText } = render(<ChannelCallout {...mockData} />);
    const linkElement = getByText('Test Link Text');
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute('href', '/en/test-link');
  });
});
