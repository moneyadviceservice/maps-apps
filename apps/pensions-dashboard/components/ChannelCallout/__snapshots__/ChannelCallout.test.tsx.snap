// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelCallout renders correctly 1`] = `
<div>
  <div
    class="relative min-h-[95px] rounded-bl-3xl before:content-[""] before:block before:h-[80px] before:w-[8px] before:absolute before:left-0 before:top-0 before:rounded-bl-[8px] bg-green-100 before:bg-green-700 p-8 pt-6 mb-12"
    data-testid="callout-positive"
  >
    <h4
      class="text-xl md:text-2xl text-gray-800 font-bold mb-3"
    >
      Test Title
    </h4>
    <p
      class="mb-4"
      data-testid="paragraph"
    >
      Test Description
    </p>
    <ul>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension Admin Highland
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension For Everyone
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        State Pension
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension Admin Highland
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension Admin Highland
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension Admin Highland
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension For Everyone
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension For Everyone
      </li>
      <li
        class="flex items-center gap-3 mt-4"
      >
        <span
          class="w-[24px] h-[24px]"
        >
          <svgr-mock
            classname="test-icon-class"
            data-testid="icon"
          />
        </span>
        Pension For Everyone
      </li>
    </ul>
    <a
      class="border-0 text-white bg-pink-600 hover:bg-pink-800 hover:shadow-none active:bg-pink-900 active:outline-yellow-200 active:text-white t-button px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-500 disabled:shadow-none focus:bg-yellow-200 focus:outline-purple-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none active:shadow-none active:outline-[3px] active:outline-offset-0 mt-6 mb-2"
      href="/en/test-link"
    >
      Test Link Text
    </a>
  </div>
</div>
`;
